### PR TITLE
Use generics instead of trait objects in FdtReader

### PR DESCRIPTION
**Pull request by doriath**
ref #38

Generics allow us to avoid dynamic dispatch and should make the code
faster.

Signed-off-by: Tomasz Zurkowski <zurkowski@google.com>

### CODE REVIEW

File: src/lib/device_tree/src/lib.rs [hunk 1]
```diff
@@ -227,7 +227,7 @@ impl<'a> Iterator for FdtIterator<'a> {
                     let value = SectionReader::new(&self.dt.struct_block, self.cursor, len);
                     self.cursor = align4(self.cursor + len);
                     // Note: This performs a copy of the whole path_buf array!
-                    return Some(Entry::Property::<'a> { path: Path { depth: self.depth + 1, len: self.len_buf, buf: self.path_buf }, value: value });
+                    return Some(Entry::Property { path: Path { depth: self.depth + 1, len: self.len_buf, buf: self.path_buf }, value: value });
```
Comment (fotonick) [inline src/lib/device_tree/src/lib.rs:230]: Clippy suggested replacing `value: value` with simply `value` in the past, but I didn't want to make a PR just for that. Could you update that here?

File: src/lib/device_tree/src/lib.rs [hunk 2]
```diff
@@ -109,10 +109,10 @@ impl Path {
...
-    _mem_reservation_block: SectionReader<'a>,
-    struct_block: SectionReader<'a>,
-    strings_block: SectionReader<'a>,
+pub struct FdtReader<'a, D: Driver> {
+    _mem_reservation_block: SectionReader<'a, D>,
+    struct_block: SectionReader<'a, D>,
+    strings_block: SectionReader<'a, D>,
 }
 
 fn cursor_u32(drv: &dyn Driver, cursor: &mut usize) -> Result<u32> {
```
Comment (fotonick) [inline src/lib/device_tree/src/lib.rs:118]: Would it be in the spirit of your PR to change all dyn to impl?

**Review by fotonick — COMMENTED** [review]
I think that this is a good idea. It doesn't seem like there will be terribly many `Driver` impls in any given build. That said, could you characterize the difference in binary size?

File: src/lib/device_tree/src/lib.rs (hunk 1 above)
Comment (doriath) [reply src/lib/device_tree/src/lib.rs:230]: I created https://github.com/oreboot/oreboot/issues/282 to do it as separate PR (fix all existing lint errors and configure CI). Is it ok to not do this here?

File: src/lib/device_tree/src/lib.rs (hunk 2 above)
Comment (doriath) [reply src/lib/device_tree/src/lib.rs:118]: Yes, changed all `dyn` uses inside device_tree crate to `impl`, thanks!

**Comment by doriath** [comment]
The size of bootblob.bin files before and after (in bytes):

| Mainboard            | Before | After | Diff |
|----------------------|--------|-------|------|
| emulation/qemu-q35   | 65536  | 65536 | 0    |
| emulation/qemu-riscv | 13288  | 13288 | 0    |
| opentitan/crb        | 7664   | 7314  | -350 |
| sifive/hifive        | 36840  | 36338 | -502 |

I have no idea why the version with generics is smaller.

Numbers were generated with:

```
git checkout 366452794cf524f475b46a752a3b397940d880f3 && make mainboards && ~/bootblob-size.sh
git checkout b16c860f0a7863abcdb69e1e30e5097e0339555a && make mainboards && ~/bootblob-size.sh
```

Where bootblob-size.sh has following content:

```bash
#!/usr/bin/env bash

for BOOTBLOB in $(find . -name bootblob.bin)
do
  SIZE=$(stat -c"%s" $BOOTBLOB)
  if [[ $BOOTBLOB =~ \./[^/]+/[^/]+/([^/]+)/([^/]+).* ]]; then 
    echo "${BASH_REMATCH[1]}/${BASH_REMATCH[2]};${SIZE}"
  fi 
done
```

File: src/lib/device_tree/src/lib.rs (hunk 1 above)
Comment (fotonick) [reply src/lib/device_tree/src/lib.rs:230]: Thanks for filing #282! I like your more systematic plan. I think it's OK to do the `value` replacement here or for addressing #282.

**Comment by fotonick** [comment]
Thank you for running the comparison!

The results are curious, though, with some sizes changing and not others, on top of the non-zero changes being negative rather than positive. I recently noticed that `flashlayout` first reserves a fixed amount of space for each area, fills with 0xFF, then overwrites with the requested file contents. There's an explicit check that errors out if the data in the file is bigger than the space reserved. This suggests to me that the binary should always be fixed size, just padded with a variable number of 0xFF.

@rjoleary, do you have any idea how the boot blobs can change size? Maybe it's not used how I think it is?

**Comment by fotonick** [comment]
Ok, figured out why I'm wrong—`bootblob.bin` is an input to `layoutflash`, not an output.

Now there's just the mystery of why no size change or shrinkage.

**Review by fotonick — APPROVED** [review]
But I realize that I don't even care about the negatives; it's the zeros that are confusing. Regardless, this PR is clearly a win.

**Pull request merged**